llvm/lib/Target/AIE/AIEBaseAliasAnalysis.cpp

```diff
@@ -117,13 +138,181 @@ static const Value *getUnderlyingObjectAIE(const Value *V) {
   return V;
 }
 
+SmallPtrSet<Value *, 5> collectIntrinsicOperands(const Value *V) {
+  const auto *Extract = dyn_cast<ExtractValueInst>(V);
+  assert(Extract && "Not an Extract");
+  const auto *IntrinsicCall =
+      dyn_cast<CallBase>(Extract->getAggregateOperand());
+  assert(IntrinsicCall && "Not an Intrinsic");
+  Intrinsic::ID ID = IntrinsicCall->getIntrinsicID();
+  SmallPtrSet<Value *, 5> ArgSet;
+  unsigned InPtrIdx;
+  isAIEPtrAddIntrinsic(ID, InPtrIdx);
+  // Necessary operands to disambiguate.
+  for (unsigned Op : getAddIntrinsicOps(ID)) {
+    ArgSet.insert(IntrinsicCall->getArgOperand(Op));
+  }
+  return ArgSet;
+}
+
+/// This function returns some tracking information related to the upward path
+/// that starts from a Value *V to the Phi node that firstly defines a pointer
+/// value to the same ambiguous memory. The returned information can be
+/// described as follows:
+/// * int: the number of times this pointer is updated through a chain of
+///   add_2d/3d intrinsics calls. If these the pointer is updated in a non
+///   uniform way, ie. using a different addressing information (offset,
+///   (increment and size) this value will be -1. We also return {0 and {}}
+///   int case we track a value that represents the external Phi incoming
+///   value.
+/// * SmallPtrSet<Value*>: represent the constant (across updates) values
+///   that are used to update the pointer across different updates.
+static std::pair<int, SmallPtrSet<Value *, 5>>
+trackIntrinsicChain(const Value *V) {
```

Comment:
I think the function would gain in readability if we had an interface like
```
struct IntrinsicChainInfo {
  const Value *BasePtr = nullptr;
  unsigned IntrinsicCalls = 0;
  SmallVector<const Value *, 5> IntrinsicParameters;
}

std::optional<IntrinsicChainInfo> trackIntrinsicChain(const Value *V);
```
It would return std::nullopt when the chain of intrinsics could not be safely followed. I.e. we replace `return {-1, {}}` by `return {}`;

We can also then check `Result->BasePtr` against `PhiA`/`PhiB`.

Reply:
Basically, we get this `PhiA` and `PhiB` by calling the calling getUnderlyingObjectAIE(A/B). Later on, we reconstruct  the tracking information from A/B to the underlying object (we redo the getUnderlyingObjectAIE collecting the information along the way). In this way,  BasePtr will always be PhiA or PhiB.